Use the parsed virtual fact provided by 1.24

ekohl opened the pull request:
Foreman 1.24 introduced an extended the reported data facets with whether a host is virtual. This is more reliable than the provider since not everyone is using the compute resource integration.

timogoebel commented:
Thanks, but doesn‘t this require that there is a report available. I suspect a timing problem because the monitoring object is being created before the first report came in.

ekohl commented:
Probably. In my case the hosts are created by the facts that come in so this works better for me. I was considering seeding if it responds to it virtual to also remain compatible with older versions. Is there a way to see if there's a fact available for it? Perhaps checking for nil?

ekohl commented:
To resolve this, should we use `virtual || provider != 'BareMetal'` then?

timogoebel commented:
I'm wondering if we should explicitly check if virtual is set. (if virtual is false because this is a bare metal host, it would still fallback to checking the provider)

`virtual.present? ? virtual : provider != 'BareMetal' `

ekohl commented:
I considered that and think you're right. Updated to use that. I think that `.present?` also makes it safe for versions prior to 1.24 so I dropped that. Please verify this is true :)

timogoebel commented:
@kamils-iRonin: Can we migrate this to github actions as well? Then we see if tests pass for all supported core versions.

kamils-iRonin commented:
> Can we migrate this to github actions as well? Then we see if tests pass for all supported core versions.

@timogoebel sure, I'll open PR

timogoebel commented:
@ekohl: Do you mind rebasing this PR? Thanks.

ekohl commented:
Rebased

ekohl commented:
At the suggestion of Rubocop I changed it to `virtual.presence` but that fails in versions prior to 1.24. Would it be preferable to raise the min version to 1.24 or use `virtual&.presence`?

timogoebel commented:
@ekohl: I think `virtual&.presence` should be fine here.

ekohl commented:
Updated though now I'm wondering if it should be rebased on #73 since that avoids dealing with older versions.

ekohl commented:
Looks like `virtual&.presence` also doesn't work.

kamils-iRonin commented:
> Looks like `virtual&.presence` also doesn't work.

in this case `virtual.try(:presence)` should be fine, or you can detect Foreman's version with `Gem::Version.new(SETTINGS[:version].notag) >= Gem::Version.new('1.24')` but it doesn't look pretty

timogoebel commented:
> I'm wondering if it should be rebased on #73 since that avoids dealing with older versions.

I've just merged #73. Let's rebase this PR on top. :)

ekohl commented:
Rebased and green now.

timogoebel commented:
Thanks!
